src/passes/Poppify.cpp

```diff
@@ -0,0 +1,504 @@
+  }
+  // If the extracted value is the only one left, we're done
+  if (curr->index == 0) {
+    return;
+  }
+  // Otherwise, save it to a scratch local and drop the other values
+  auto type = types[curr->index];
+  Index scratch = getScratchLocal(type);
+  instrs.push_back(builder.makeLocalSet(scratch, builder.makePop(type)));
+  for (size_t i = curr->index - 1; i != size_t(-1); --i) {
+    instrs.push_back(builder.makeDrop(builder.makePop(types[i])));
+  }
+  // Retrieve the saved value
+  instrs.push_back(builder.makeLocalGet(scratch, type));
+}
+
+void Poppifier::emitDrop(Drop* curr) {
+  auto& instrs = scopeStack.back().instrs;
+  if (curr->value->type.isTuple()) {
+    // Drop each element individually
+    auto types = curr->value->type;
+    for (auto it = types.rbegin(), end = types.rend(); it != end; ++it) {
+      instrs.push_back(builder.makeDrop(builder.makePop(*it)));
+    }
+  } else {
+    poppify(curr);
+    instrs.push_back(curr);
+  }
+}
+
+void Poppifier::emitLocalGet(LocalGet* curr) {
+  auto& instrs = scopeStack.back().instrs;
+  if (curr->type.isTuple()) {
+    auto types = func->getLocalType(curr->index);
+    const auto& elems = tupleVars[curr->index];
+    for (size_t i = 0; i < types.size(); ++i) {
+      instrs.push_back(builder.makeLocalGet(elems[i], types[i]));
+    }
+  } else {
+    instrs.push_back(curr);
+  }
+}
+
+void Poppifier::emitLocalSet(LocalSet* curr) {
+  auto& instrs = scopeStack.back().instrs;
+  if (curr->value->type.isTuple()) {
+    auto types = func->getLocalType(curr->index);
+    const auto& elems = tupleVars[curr->index];
+    // Add the unconditional sets
+    for (size_t i = types.size() - 1; i >= 1; --i) {
+      instrs.push_back(
+        builder.makeLocalSet(elems[i], builder.makePop(types[i])));
+    }
+    if (curr->isTee()) {
+      // Use a tee followed by gets to retrieve the tuple
+      instrs.push_back(
+        builder.makeLocalTee(elems[0], builder.makePop(types[0]), types[0]));
+      for (size_t i = 1; i < types.size(); ++i) {
+        instrs.push_back(builder.makeLocalGet(elems[i], types[i]));
+      }
+    } else {
+      // Otherwise just add the last set
+      instrs.push_back(
+        builder.makeLocalSet(elems[0], builder.makePop(types[0])));
+    }
+  } else {
+    poppify(curr);
+    instrs.push_back(curr);
+  }
+}
+
+void Poppifier::emitGlobalGet(GlobalGet* curr) {
+  auto& instrs = scopeStack.back().instrs;
+  if (curr->type.isTuple()) {
+    auto types = module->getGlobal(curr->name)->type;
+    for (Index i = 0; i < types.size(); ++i) {
+      instrs.push_back(
+        builder.makeGlobalGet(getGlobalElem(curr->name, i), types[i]));
+    }
+  } else {
+    instrs.push_back(curr);
+  }
+}
+
+void Poppifier::emitGlobalSet(GlobalSet* curr) {
+  auto& instrs = scopeStack.back().instrs;
+  if (curr->value->type.isTuple()) {
+    auto types = module->getGlobal(curr->name)->type;
+    for (Index i = types.size(); i > 0; --i) {
```

Comment:
Nit: Do we need to iterate in the reverse order? (`emitLocalSet` does though)

Reply:
Yes. Consider this:

```
(global.set $tuple
  (tuple.make
     (i32.const 0)
     (i32.const 1)
     (i32.const 2)
  )
)
```

The `tuple.make` is eliminated by Poppification, so we end up with 0, 1, and 2 on the stack. Since the 2 is at the top of stack, it needs to be consumed first, so we need to work through the values from last pushed to first pushed.